src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Logging/MessageOrigin.cs

```diff
@@ -56,24 +58,38 @@ public MessageOrigin(string fileName, int sourceLine, int sourceColumn, ModuleDe
 
         public MessageOrigin(MethodIL methodBody, int ilOffset)
         {
-            string? document = null;
-            int? lineNumber = null;
-
+            ILSequencePoint? correspondingSequencePoint = null;
             IEnumerable<ILSequencePoint>? sequencePoints = methodBody.GetDebugInfo()?.GetSequencePoints();
             if (sequencePoints != null)
             {
                 foreach (var sequencePoint in sequencePoints)
                 {
                     if (sequencePoint.Offset <= ilOffset)
                     {
-                        document = sequencePoint.Document;
-                        lineNumber = sequencePoint.LineNumber;
+                        correspondingSequencePoint = sequencePoint;
+                    }
+                }
+
+                // If the warning comes from hidden line (compiler generated code typically)
+                // search for any sequence point with non-hidden line number and report that as a best effort.
```

Comment:
In a way doing that can lead to more confusion. If the warning is in the compiler generated portion of the method, then "assigning" it to something "close by" is as wrong as assigning it to pretty much any other part of the method - mainly because we have no idea how the method is split between compiler generated and user code. So pointing to the method as a whole feels like a good compromise.

If we do find a case where this would improve experience I'm not against doing it, in all tools at once.